common/djangoapps/entitlements/models.py

```diff
@@ -0,0 +1,98 @@
+from django.db import models
+from django.contrib.auth.models import User
+from course_modes.models import CourseMode
+
+
+class CourseEntitlement(models.Model):
+    """
+    Represents a Student's Entitlement to a Course Run for a given Course.
+    """
+
+    user = models.ForeignKey(User)
+
+    # The parent courses UUID
+    course_uuid = models.UUIDField()
+
+    created = models.DateTimeField(auto_now_add=True, db_index=True)
+
+    updated = models.DateTimeField(auto_now=True)
+
+    # The date that an the entitlement expired
+    # if NULL the entitlement has not expired
+    expired_at = models.DateTimeField(null=True)
+
+    # The mode of the Course that will be applied
+    mode = models.CharField(default=CourseMode.DEFAULT_MODE_SLUG, max_length=100)
+
+    # The ID of the course enrollment for this Entitlement
+    # if NULL the entitlement is not in use
+    enrollment_course_run = models.ForeignKey('student.CourseEnrollment', null=True)
+
+    @classmethod
+    def entitlements_for_user(cls, user):
+        """
+        Retrieve all the Entitlements for a User
+
+        Arguments:
+            user: A Django User object identifying the current user
+
+        Returns:
+            All of the Entitlements for the User
+        """
+        return cls.objects.filter(user=user)
+
+    @classmethod
+    def get_user_course_entitlement(cls, user, course_uuid):
+        """
+        Retrieve The entitlement for the given parent course id if it exists for the User
+
+        Arguments:
+            user: A Django User object identifying the current user
+            course_uuid(string): The parent course uuid
+
+        Returns:
+            The single entitlement for the requested parent course id
+        """
+        return cls.objects.filter(user=user, course_uuid=course_uuid).first()
+
+    @classmethod
+    def update_or_create_new_entitlement(cls, user, course_uuid, entitlement_data):
```

Comment:
I'm a little confused as to why we need this method.

Reply:
Used in testing here, but also used in the next PR for the API.  I prefer to wrap the query for when to update/create a new entitlement as to not duplicate the query logic.

Reply:
as long as there is actual logic that will be applied this is fine.  It doesn't make sense to wrap this as written just to avoid writing `.objects.`.  But if we are planning on adding logic to this method later then 👍